compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -169,69 +243,247 @@ data UpdChoice = UpdChoice
+  data Env ph
+  -- | Construct a base update.
+  baseUpd :: Rec [Cond BaseUpd] -> Upd ph
+  -- | Construct a choice exercise update.
+  choiceUpd :: Cond UpdChoice -> Upd ph
+  -- | Construct a value update.
+  valueUpd :: Cond (Qualified ExprValName) -> Upd ph
+  -- | Map over a single base update.
+  mapBaseUpd :: (Rec [Cond BaseUpd] -> Upd ph) -> Upd ph -> Upd ph
+  -- | Check whether the update set contains any choice references.
+  containsChoiceRefs :: UpdateSet ph -> Bool
+  -- | Extend the conditional of an update.
+  extCondUpd :: BoolExpr -> Upd ph -> UpdateSet ph
+  -- | Construct an empty environment.
   emptyEnv :: Env ph
-
-instance GenPhase 'ValueGathering where
-  emptyUpdateSet = UpdateSetVG [] [] []
+  -- | Combine two environments.
+  concatEnv :: Env ph -> Env ph -> Env ph
+  -- | Get the skolemised term variables and fields from the environment.
+  envSkols :: Env ph -> [Skolem]
+  -- | Update the skolemised term variables and fields in the environment.
+  setEnvSkols :: [Skolem] -> Env ph -> Env ph
+  -- | Get the bound values from the environment.
+  envVals :: Env ph -> HM.HashMap (Qualified ExprValName) (Expr, UpdateSet ph)
+  -- | Update the bound values in the environment.
+  setEnvVals :: HM.HashMap (Qualified ExprValName) (Expr, UpdateSet ph) -> Env ph -> Env ph
+  -- | Get the data constructors from the environment.
+  envDats :: Env ph -> HM.HashMap TypeConName DefDataType
+  -- | Update the data constructors in the environment.
+  setEnvDats :: HM.HashMap TypeConName DefDataType -> Env ph -> Env ph
+  -- | Get the fetched cid's mapped to their current variable name, along with
+  -- a list of any potential old variable names, from the environment.
+  envCids :: Env ph -> HM.HashMap Cid (ExprVarName, [ExprVarName])
+  -- | Update the fetched cid's in the environment.
+  setEnvCids :: HM.HashMap Cid (ExprVarName, [ExprVarName]) -> Env ph -> Env ph
+  -- | Get the set of preconditions from the environment.
+  envPreconds :: Env ph -> HM.HashMap (Qualified TypeConName) (Expr -> Expr)
+  -- | Update the set of preconditions in the environment.
+  setEnvPreconds :: HM.HashMap (Qualified TypeConName) (Expr -> Expr) -> Env ph -> Env ph
+  -- | Get the additional constraints from the environment.
+  envCtrs :: Env ph -> [BoolExpr]
+  -- | Update the additional constraints in the environment.
+  setEnvCtrs :: [BoolExpr] -> Env ph -> Env ph
+  -- | Get the set of relevant choices from the environment.
+  envChoices :: Env ph -> HM.HashMap UpdChoice (ChoiceData ph)
+  -- | Update the set of relevant choices in the environment.
+  setEnvChoices :: HM.HashMap UpdChoice (ChoiceData ph) -> Env ph -> Env ph
+
+instance IsPhase 'ValueGathering where
+  data Upd 'ValueGathering
+    = UpdVGBase ![Cond BaseUpd]
+    -- ^ A base update.
+    | UpdVGChoice !(Cond UpdChoice)
+    -- ^ An exercised choice.
+    | UpdVGVal !(Cond (Qualified ExprValName))
+    -- ^ A referenced value.
+  data Env 'ValueGathering = EnvVG
+    ![Skolem]
```

Comment:
At least the 4 fields seem to be common to all 3 `Env` types and there also seem to be shared parts between the latter fields in particular for `ChoiceGathering` and `Solving`. How about we either factor them out or factor out the whole `Env` type?

Reply:
What exactly do you mean with `factor out` here? I very much dislike the current definition of `Env`, because there is a lot of overlap indeed. But I don't see a nicer solution?

Reply:
I see two solutions:
1. Use a single `Env` type with one or more type parameters for the fields that are different between the 3 phases. 1 + type families should be enough but you can argue whether type families are easier or harder to understand here compared to just having separate type parameters.
2. Keep using different `Env` types but factor out the shared parts into separate types, e.g., you could make a type with 4 fields that captures the 4 fields currently in `Env` and then give the `Env` types a single field that references that type.

I’d probably go for 1 since you then have a single Env type that you can operate on generically instead of needing all those `get` methods for things that are common to all 3.

Reply:
Cool! I guess that would work indeed, and it would certainly reduce the duplication here :)